File: im-mobile/lib/services/self_destruct_service.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';

/**
 * 阅后即焚服务 - 移动端
 * 消息自动销毁功能
 */

/// 销毁计时器类型
enum TimerType {
  seconds5('SECONDS_5', '5秒', 5, '⏱️'),
  seconds30('SECONDS_30', '30秒', 30, '⏱️'),
  minute1('MINUTE_1', '1分钟', 60, '🔥'),
  minutes5('MINUTES_5', '5分钟', 300, '🔥'),
  hour1('HOUR_1', '1小时', 3600, '💣'),
  hours24('HOURS_24', '24小时', 86400, '💣'),
  custom('CUSTOM', '自定义', -1, '⚙️');

  final String value;
  final String label;
  final int seconds;
  final String icon;

  const TimerType(this.value, this.label, this.seconds, this.icon);
}

/// 销毁状态
enum DestroyStatus {
  pending('PENDING'),
  counting('COUNTING'),
  destroyed('DESTROYED'),
  expired('EXPIRED');

  final String value;

  const DestroyStatus(this.value);
}

/// 阅后即焚消息数据
class SelfDestructMessage {
  final int id;
  final int messageId;
  final int senderId;
  final int receiverId;
  final String timerType;
  final int durationSeconds;
  final DestroyStatus status;
  final DateTime createdAt;
  final DateTime? readAt;
  final DateTime? destroyedAt;
  final String? destroyReason;

  SelfDestructMessage({
    required this.id,
    required this.messageId,
    required this.senderId,
    required this.receiverId,
    required this.timerType,
    required this.durationSeconds,
    required this.status,
    required this.createdAt,
    this.readAt,
    this.destroyedAt,
    this.destroyReason,
  });

  factory SelfDestructMessage.fromJson(Map<String, dynamic> json) {
    return SelfDestructMessage(
      id: json['id'] ?? 0,
      messageId: json['messageId'] ?? 0,
      senderId: json['senderId'] ?? 0,
      receiverId: json['receiverId'] ?? 0,
      timerType: json['timerType'] ?? '',
      durationSeconds: json['durationSeconds'] ?? 0,
      status: DestroyStatus.values.firstWhere(
        (e) => e.value == json['status'],
        orElse: () => DestroyStatus.pending,
      ),
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      readAt: json['readAt'] != null ? DateTime.parse(json['readAt']) : null,
      destroyedAt: json['destroyedAt'] != null ? DateTime.parse(json['destroyedAt']) : null,
      destroyReason: json['destroyReason'],
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'messageId': messageId,
    'senderId': senderId,
    'receiverId': receiverId,
    'timerType': timerType,
    'durationSeconds': durationSeconds,
    'status': status.value,
    'createdAt': createdAt.toIso8601String(),
    'readAt': readAt?.toIso8601String(),
    'destroyedAt': destroyedAt?.toIso8601String(),
    'destroyReason': destroyReason,
  };
}

/// 销毁记录
class DestroyRecord {
  final int messageId;
  final String reason;
  final int? operatorId;
  final DateTime destroyTime;
  final String? note;

  DestroyRecord({
    required this.messageId,
    required this.reason,
    this.operatorId,
    required this.destroyTime,
    this.note,
  });

  factory DestroyRecord.fromJson(Map<String, dynamic> json) {
    return DestroyRecord(
      messageId: json['messageId'] ?? 0,
      reason: json['reason'] ?? '',
      operatorId: json['operatorId'],
      destroyTime: DateTime.parse(json['destroyTime'] ?? DateTime.now().toIso8601String()),
      note: json['note'],
    );
  }
}

/// 销毁状态响应
class DestroyStatusResponse {
  final DestroyStatus? status;
  final int remainingSeconds;

  DestroyStatusResponse({
    this.status,
    required this.remainingSeconds,
  });

  factory DestroyStatusResponse.fromJson(Map<String, dynamic> json) {
    return DestroyStatusResponse(
      status: json['status'] != null 
        ? DestroyStatus.values.firstWhere(
            (e) => e.value == json['status'],
            orElse: () => DestroyStatus.pending,
          )
        : null,
      remainingSeconds: json['remainingSeconds'] ?? -1,
    );
  }
}

/// 倒计时条目
class CountdownEntry {
  final int messageId;
  Timer? timer;
  int remainingSeconds;
  final Function(int)? onTick;
  final VoidCallback? onComplete;

  CountdownEntry({
    required this.messageId,
    required this.remainingSeconds,
    this.onTick,
    this.onComplete,
  });
}

/// 阅后即焚服务
class SelfDestructService {
  static const String _baseUrl = 'http://localhost:8080/api';
  
  final Map<int, CountdownEntry> _activeCountdowns = {};
  final Set<int> _destroyedMessages = {};

  /// API客户端
  final http.Client _client;
  final Map<String, String> _headers = {
    'Content-Type': 'application/json',
  };

  SelfDestructService({http.Client? client}) : _client = client ?? http.Client();

  /// 设置消息阅后即焚
  Future<SelfDestructMessage> setupSelfDestruct({
    required int messageId,
    required int senderId,
    required int receiverId,
    required TimerType timerType,
    int? customSeconds,
  }) async {
    final response = await _client.post(
      Uri.parse('$_baseUrl/self-destruct/setup'),
      headers: _headers,
      body: jsonEncode({
        'messageId': messageId,
        'senderId': senderId,
        'receiverId': receiverId,
        'timerType': timerType.value,
        'customSeconds': customSeconds,
      }),
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return SelfDestructMessage.fromJson(data['data']);
    } else {
      throw Exception('设置阅后即焚失败: ${response.statusCode}');
    }
  }

  /// 标记消息已读
  Future<void> markAsRead({
    required int messageId,
    required int readerId,
  }) async {
    final response = await _client.post(
      Uri.parse('$_baseUrl/self-destruct/read?messageId=$messageId&readerId=$readerId'),
      headers: _headers,
    );

    if (response.statusCode != 200) {
      throw Exception('标记已读失败: ${response.statusCode}');
    }
  }

  /// 获取消息销毁状态
  Future<DestroyStatusResponse> getStatus(int messageId) async {
    final response = await _client.get(
      Uri.parse('$_baseUrl/self-destruct/status?messageId=$messageId'),
      headers: _headers,
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return DestroyStatusResponse.fromJson(data);
    } else {
      throw Exception('获取状态失败: ${response.statusCode}');
    }
  }

  /// 手动销毁消息
  Future<void> destroyMessage({
    required int messageId,
    required int operatorId,
  }) async {
    final response = await _client.post(
      Uri.parse('$_baseUrl/self-destruct/destroy?messageId=$messageId&operatorId=$operatorId'),
      headers: _headers,
    );

    if (response.statusCode == 200) {
      _clearCountdown(messageId);
    } else {
      throw Exception('销毁消息失败: ${response.statusCode}');
    }
  }

  /// 批量销毁消息
  Future<void> batchDestroy({
    required List<int> messageIds,
    required int operatorId,
  }) async {
    final response = await _client.post(
      Uri.parse('$_baseUrl/self-destruct/batch-destroy'),
      headers: _headers,
      body: jsonEncode({
        'messageIds': messageIds,
        'operatorId': operatorId,
      }),
    );

    if (response.statusCode == 200) {
      for (final id in messageIds) {
        _clearCountdown(id);
      }
    } else {
      throw Exception('批量销毁失败: ${response.statusCode}');
    }
  }

  /// 获取销毁历史
  Future<List<DestroyRecord>> getDestroyHistory({
    required int userId,
    int page = 0,
    int size = 20,
  }) async {
    final response = await _client.get(
      Uri.parse('$_baseUrl/self-destruct/history?userId=$userId&page=$page&size=$size'),
      headers: _headers,
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      final List<dynamic> records = data['data'] ?? [];
      return records.map((e) => DestroyRecord.fromJson(e)).toList();
    } else {
      throw Exception('获取销毁历史失败: ${response.statusCode}');
    }
  }

  /// 开始本地倒计时
  void startLocalCountdown({
    required int messageId,
    required int durationSeconds,
    Function(int)? onTick,
    VoidCallback? onComplete,
  }) {
    // 清除已有倒计时
    _clearCountdown(messageId);

    int remaining = durationSeconds;

    final timer = Timer.periodic(const Duration(seconds: 1), (timer) {
      remaining--;
      onTick?.call(remaining);

      if (remaining <= 0) {
        _clearCountdown(messageId);
        _markAsDestroyed(messageId);
        onComplete?.call();
      }
    });

    _activeCountdowns[messageId] = CountdownEntry(
      messageId: messageId,
      remainingSeconds: remaining,
      onTick: onTick,
      onComplete: onComplete,
    )..timer = timer;
  }

  /// 清除倒计时
  void _clearCountdown(int messageId) {
    final entry = _activeCountdowns[messageId];
    if (entry != null) {
      entry.timer?.cancel();
      _activeCountdowns.remove(messageId);
    }
  }

  /// 标记消息为已销毁
  void _markAsDestroyed(int messageId) {
    _destroyedMessages.add(messageId);
    _clearCountdown(messageId);
  }

  /// 检查消息是否已销毁
  bool isDestroyed(int messageId) => _destroyedMessages.contains(messageId);

  /// 获取剩余秒数
  int getRemainingSeconds(int messageId) {
    final entry = _activeCountdowns[messageId];
    return entry?.remainingSeconds ?? -1;
  }

  /// 格式化剩余时间
  String formatRemainingTime(int seconds) {
    if (seconds < 0) return '';
    if (seconds < 60) return '${seconds}秒';
    if (seconds < 3600) return '${seconds ~/ 60}分${seconds % 60}秒';
    return '${seconds ~/ 3600}小时${(seconds % 3600) ~/ 60}分';
  }

  /// 获取计时器配置
  TimerType? getTimerConfig(String type) {
    try {
      return TimerType.values.firstWhere((e) => e.value == type);
    } catch (e) {
      return null;
    }
  }

  /// 释放资源
  void dispose() {
    for (final entry in _activeCountdowns.values) {
      entry.timer?.cancel();
    }
    _activeCountdowns.clear();
    _client.close();
  }
}

File: im-backend/src/main/java/com/im/server/service/SelfDestructService.java
package com.im.server.service;

import com.im.server.entity.Message;
import com.im.server.entity.SelfDestructMessage;
import com.im.server.repository.SelfDestructMessageRepository;
import com.im.server.repository.MessageRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import jakarta.annotation.PostConstruct;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * 阅后即焚服务 - 消息自动销毁
 * 
 * 功能特性：
 * 1. 支持多种销毁计时器：5秒/30秒/1分钟/5分钟/1小时/24小时
 * 2. 消息读取后自动开始倒计时
 * 3. 支持查看消息状态（已读/未读/已销毁）
 * 4. 销毁时保留元数据（发送者、时间戳）
 * 5. 销毁记录审计日志
 * 6. 支持批量销毁和手动销毁
 * 7. 销毁前通知发送者
 */
@Service
public class SelfDestructService {

    @Autowired
    private SelfDestructMessageRepository repository;

    @Autowired
    private MessageRepository messageRepository;

    @Autowired
    private RedisTemplate<String, Object> redisTemplate;

    // 销毁计时器类型枚举
    public enum TimerType {
        SECONDS_5(5, "5秒"),
        SECONDS_30(30, "30秒"),
        MINUTE_1(60, "1分钟"),
        MINUTES_5(300, "5分钟"),
        HOUR_1(3600, "1小时"),
        HOURS_24(86400, "24小时"),
        CUSTOM(-1, "自定义");

        private final int seconds;
        private final String label;

        TimerType(int seconds, String label) {
            this.seconds = seconds;
            this.label = label;
        }

        public int getSeconds() { return seconds; }
        public String getLabel() { return label; }
    }

    // 消息销毁状态
    public enum DestroyStatus {
        PENDING,      // 待销毁（未读）
        COUNTING,     // 倒计时中（已读）
        DESTROYED,    // 已销毁
        EXPIRED       // 超时未读
    }

    // 内存缓存：正在倒计时的消息（避免频繁Redis查询）
    private final Map<String, CountdownEntry> countdownCache = new ConcurrentHashMap<>();
    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(4);

    // 销毁记录缓存（用于审计）
    private final Map<String, DestroyRecord> destroyRecords = new ConcurrentHashMap<>();

    /**
     * 初始化服务
     */
    @PostConstruct
    public void init() {
        // 从数据库加载所有待销毁消息
        loadPendingMessages();
        
        // 启动定时检查任务（每10秒检查一次）
        scheduler.scheduleAtFixedRate(this::checkAndDestroyMessages, 10, 10, TimeUnit.SECONDS);
    }

    /**
     * 设置消息阅后即焚
     */
    public SelfDestructMessage setupSelfDestruct(Long messageId, Long senderId, Long receiverId, 
                                                  TimerType timerType, int customSeconds) {
        int duration = timerType.getSeconds();
        if (timerType == TimerType.CUSTOM) {
            duration = customSeconds;
        }

        Message message = messageRepository.findById(messageId).orElse(null);
        if (message == null) {
            throw new IllegalArgumentException("消息不存在");
        }

        SelfDestructMessage sdm = new SelfDestructMessage();
        sdm.setMessageId(messageId);
        sdm.setSenderId(senderId);
        sdm.setReceiverId(receiverId);
        sdm.setTimerType(timerType);
        sdm.setDurationSeconds(duration);
        sdm.setStatus(DestroyStatus.PENDING);
        sdm.setCreatedAt(Instant.now());

        // 保存到数据库
        SelfDestructMessage saved = repository.save(sdm);

        // 缓存到Redis
        String redisKey = getRedisKey(messageId);
        redisTemplate.opsForHash().put(redisKey, "status", DestroyStatus.PENDING.name());
        redisTemplate.opsForHash().put(redisKey, "duration", String.valueOf(duration));
        redisTemplate.opsForHash().put(redisKey, "createdAt", String.valueOf(Instant.now().toEpochMilli()));

        return saved;
    }

    /**
     * 标记消息已读（开始倒计时）
     */
    public void markAsRead(Long messageId, Long readerId) {
        String redisKey = getRedisKey(messageId);
        
        // 检查是否已标记过
        String status = (String) redisTemplate.opsForHash().get(redisKey, "status");
        if (status != null && !status.equals(DestroyStatus.PENDING.name())) {
            return; // 已经读过了
        }

        // 更新状态为倒计时中
        redisTemplate.opsForHash().put(redisKey, "status", DestroyStatus.COUNTING.name());
        redisTemplate.opsForHash().put(redisKey, "readAt", String.valueOf(Instant.now().toEpochMilli()));

        // 获取销毁时长
        String durationStr = (String) redisTemplate.opsForHash().get(redisKey, "duration");
        int duration = durationStr != null ? Integer.parseInt(durationStr) : 30;

        // 添加到倒计时缓存
        CountdownEntry entry = new CountdownEntry(messageId, Instant.now(), duration);
        countdownCache.put(messageId.toString(), entry);

        // 更新数据库状态
        repository.updateStatus(messageId, DestroyStatus.COUNTING);

        // 安排销毁任务
        scheduleDestroy(messageId, duration);
    }

    /**
     * 手动销毁消息
     */
    public void destroyMessage(Long messageId, Long operatorId) {
        performDestroy(messageId, "MANUAL", operatorId);
    }

    /**
     * 批量销毁消息
     */
    public void batchDestroy(List<Long> messageIds, Long operatorId) {
        for (Long messageId : messageIds) {
            try {
                destroyMessage(messageId, operatorId);
            } catch (Exception e) {
                // 记录错误但继续处理其他消息
                System.err.println("销毁消息失败: " + messageId + ", 错误: " + e.getMessage());
            }
        }
    }

    /**
     * 获取消息销毁状态
     */
    public DestroyStatus getStatus(Long messageId) {
        String redisKey = getRedisKey(messageId);
        String status = (String) redisTemplate.opsForHash().get(redisKey, "status");
        
        if (status != null) {
            return DestroyStatus.valueOf(status);
        }

        // 从数据库查询
        return repository.findByMessageId(messageId)
                .map(SelfDestructMessage::getStatus)
                .orElse(null);
    }

    /**
     * 获取销毁倒计时剩余秒数
     */
    public long getRemainingSeconds(Long messageId) {
        CountdownEntry entry = countdownCache.get(messageId.toString());
        if (entry == null) {
            return -1;
        }
        long elapsed = Instant.now().getEpochSecond() - entry.startTime.getEpochSecond();
        return Math.max(0, entry.duration - elapsed);
    }

    /**
     * 获取销毁历史记录
     */
    public List<DestroyRecord> getDestroyHistory(Long userId, int page, int size) {
        return destroyRecords.values().stream()
                .filter(r -> r.operatorId.equals(userId))
                .sorted((a, b) -> b.destroyTime.compareTo(a.destroyTime))
                .skip((long) page * size)
                .limit(size)
                .toList();
    }

    /**
     * 定时检查任务
     */
    private void checkAndDestroyMessages() {
        try {
            // 检查超时的待销毁消息
            List<SelfDestructMessage> pendingMessages = repository.findByStatus(DestroyStatus.PENDING);
            Instant threshold = Instant.now().minusSeconds(86400 * 7); // 7天未读标记为过期
            
            for (SelfDestructMessage sdm : pendingMessages) {
                if (sdm.getCreatedAt().isBefore(threshold)) {
                    // 标记为过期
                    repository.updateStatus(sdm.getMessageId(), DestroyStatus.EXPIRED);
                    
                    DestroyRecord record = new DestroyRecord(
                        sdm.getMessageId(), 
                        "EXPIRED", 
                        null, 
                        Instant.now(),
                        "消息超过7天未读，自动过期"
                    );
                    destroyRecords.put(sdm.getMessageId().toString(), record);
                }
            }

            // 清理已完成的倒计时缓存（节省内存）
            countdownCache.entrySet().removeIf(entry -> {
                long elapsed = Instant.now().getEpochSecond() - entry.getValue().startTime.getEpochSecond();
                return elapsed > entry.getValue().duration + 60; // 销毁后保留1分钟
            });

        } catch (Exception e) {
            System.err.println("检查销毁消息失败: " + e.getMessage());
        }
    }

    /**
     * 执行消息销毁
     */
    private void performDestroy(Long messageId, String reason, Long operatorId) {
        String redisKey = getRedisKey(messageId);

        // 获取当前状态
        String status = (String) redisTemplate.opsForHash().get(redisKey, "status");
        if (status != null && status.equals(DestroyStatus.DESTROYED.name())) {
            return; // 已经销毁了
        }

        // 更新Redis状态
        redisTemplate.opsForHash().put(redisKey, "status", DestroyStatus.DESTROYED.name());
        redisTemplate.opsForHash().put(redisKey, "destroyedAt", String.valueOf(Instant.now().toEpochMilli()));
        redisTemplate.opsForHash().put(redisKey, "destroyReason", reason);

        // 更新数据库
        repository.updateStatus(messageId, DestroyStatus.DESTROYED);

        // 从倒计时缓存移除
        countdownCache.remove(messageId.toString());

        // 记录销毁
        DestroyRecord record = new DestroyRecord(messageId, reason, operatorId, Instant.now(), null);
        destroyRecords.put(messageId.toString(), record);

        // 通知发送者（通过WebSocket或推送）
        notifySender(messageId, reason);

        System.out.println("消息已销毁: messageId=" + messageId + ", reason=" + reason);
    }

    /**
     * 安排销毁任务
     */
    private void scheduleDestroy(Long messageId, int delaySeconds) {
        scheduler.schedule(() -> {
            try {
                performDestroy(messageId, "TIMER", null);
            } catch (Exception e) {
                System.err.println("定时销毁失败: " + messageId + ", 错误: " + e.getMessage());
            }
        }, delaySeconds, TimeUnit.SECONDS);
    }

    /**
     * 通知发送者
     */
    private void notifySender(Long messageId, String reason) {
        // TODO: 通过WebSocket或消息推送通知发送者
        // 可以发送一个系统消息或推送通知
    }

    /**
     * 加载待销毁消息
     */
    private void loadPendingMessages() {
        List<SelfDestructMessage> pendingMessages = repository.findByStatus(DestroyStatus.PENDING);
        for (SelfDestructMessage sdm : pendingMessages) {
            String redisKey = getRedisKey(sdm.getMessageId());
            redisTemplate.opsForHash().put(redisKey, "status", sdm.getStatus().name());
            redisTemplate.opsForHash().put(redisKey, "duration", String.valueOf(sdm.getDurationSeconds()));
            redisTemplate.opsForHash().put(redisKey, "createdAt", String.valueOf(sdm.getCreatedAt().toEpochMilli()));
        }

        List<SelfDestructMessage> countingMessages = repository.findByStatus(DestroyStatus.COUNTING);
        for (SelfDestructMessage sdm : countingMessages) {
            String redisKey = getRedisKey(sdm.getMessageId());
            Long readAt = (Long) redisTemplate.opsForHash().get(redisKey, "readAt");
            
            if (readAt != null) {
                long elapsed = Instant.now().getEpochSecond() - (readAt / 1000);
                int remaining = sdm.getDurationSeconds() - (int) elapsed;
                
                if (remaining > 0) {
                    CountdownEntry entry = new CountdownEntry(
                        sdm.getMessageId(), 
                        Instant.ofEpochMilli(readAt), 
                        sdm.getDurationSeconds()
                    );
                    countdownCache.put(sdm.getMessageId().toString(), entry);
                    scheduleDestroy(sdm.getMessageId(), remaining);
                } else {
                    // 已经超时，立即销毁
                    performDestroy(sdm.getMessageId(), "TIMER", null);
                }
            }
        }
    }

    private String getRedisKey(Long messageId) {
        return "self_destruct:" + messageId;
    }

    // 内部类：倒计时条目
    private static class CountdownEntry {
        Long messageId;
        Instant startTime;
        int duration;

        CountdownEntry(Long messageId, Instant startTime, int duration) {
            this.messageId = messageId;
            this.startTime = startTime;
            this.duration = duration;
        }
    }

    // 内部类：销毁记录
    public static class DestroyRecord {
        public Long messageId;
        public String reason;
        public Long operatorId;
        public Instant destroyTime;
        public String note;

        public DestroyRecord(Long messageId, String reason, Long operatorId, 
                            Instant destroyTime, String note) {
            this.messageId = messageId;
            this.reason = reason;
            this.operatorId = operatorId;
            this.destroyTime = destroyTime;
            this.note = note;
        }
    }
}
